src/Microsoft.AspNetCore.Mvc.Core/ApplicationModels/ActionAttributeRouteModel.cs

```diff
@@ -3,11 +3,141 @@
 
 using System.Collections.Generic;
 using System.Linq;
+using Microsoft.AspNetCore.Mvc.ActionConstraints;
+using Microsoft.AspNetCore.Mvc.Routing;
 
 namespace Microsoft.AspNetCore.Mvc.ApplicationModels
 {
     internal static class ActionAttributeRouteModel
     {
+        public static IEnumerable<SelectorModel> FlattenSelectors(ActionModel actionModel)
+        {
+            // Loop through all attribute routes defined on the controller. 
+            // These perform a cross-product with all of the action-level attribute routes.
+            var controllerSelectors = actionModel.Controller.Selectors
+                .Where(sm => sm.AttributeRouteModel != null)
+                .ToList();
+
+            // We also include metadata an action constraints from the controller
+            // even when there are no routes, or when an action overrides the route template.
+            SelectorModel additionalSelector = null;
+            if (actionModel.Controller.Selectors.Count > 0)
+            {
+                // This logic seems arbitrary but there's a good reason for it.
+                // 
+                // When we build the controller level selectors, any metadata or action constraints
+                // that aren't IRouteTemplateProvider will be included in all selectors. So we 
+                // pick any selector and then grab all of the stuff that isn't IRouteTemplateProvider
+                // then we've found all of the items that aren't routes.
+                //
+                // This is fragile wrt application model customizing the data - but no one has
+                // run into an issue with this and its pretty esoteric.
```

Comment:
I've added more comments here. This area is spooky so I don't really want to try and change how it works even though it's a little wierd.